subsys/net/lib/ocpp/ocpp.c

```diff
@@ -0,0 +1,748 @@
+		}
+
+		if (*skey) {
+			enum ocpp_key key;
+
+			key = ocpp_key_to_cfg(skey);
+			ocpp_get_configuration(key, ctx, uid);
+		} else {
+			for (i = 0; i < OCPP_CFG_END; i++) {
+				ocpp_get_configuration(i, ctx, uid);
+			}
+		}
+
+		break;
+
+	case PDU_CHANGE_CONFIGURATION:
+		char sval[CISTR500] = {0};
+
+		memset(skey, 0, sizeof(skey));
+		ret = fn(ui->recv_buf, skey, sval);
+		if (ret) {
+			break;
+		}
+
+		ocpp_change_configuration(skey, ctx, sval, uid);
+		break;
+
+	case PDU_HEARTBEAT:
+		/* todo : sync time */
+		break;
+
+	case PDU_REMOTE_START_TRANSACTION:
+		memset(&msg, 0, OCPP_INTERNAL_MSG_SIZE);
+		msg.msgtype = PDU_REMOTE_START_TRANSACTION;
+
+		ret = fn(ui->recv_buf, &msg.usr.start_charge.id_con,
+			 msg.usr.start_charge.idtag);
+		if (ret) {
+			break;
+		}
+
+		ocpp_remote_start_transaction(ctx, &msg, uid);
+		break;
+
+	case PDU_REMOTE_STOP_TRANSACTION:
+		memset(&msg, 0, OCPP_INTERNAL_MSG_SIZE);
+		msg.msgtype = PDU_REMOTE_STOP_TRANSACTION;
+
+		ret = fn(ui->recv_buf, &idtxn);
+		if (ret) {
+			break;
+		}
+
+		ocpp_remote_stop_transaction(ctx, &msg, idtxn, uid);
+		break;
+
+	case PDU_UNLOCK_CONNECTOR:
+		memset(&msg, 0, OCPP_INTERNAL_MSG_SIZE);
+		msg.msgtype = PDU_UNLOCK_CONNECTOR;
+
+		ret = fn(ui->recv_buf, &msg.usr.unlock_con.id_con);
+		if (ret) {
+			break;
+		}
+
+		ocpp_unlock_connector(ctx, &msg, uid);
+		break;
+
+	case PDU_METER_VALUES:
+	default:
+		break;
+	}
+
+	if (is_rsp) {
+		k_poll_signal_raise(&ui->ws_rspsig, 0);
+	}
+	return 0;
+}
+
+#define TCP_CONNECT_AFTER 20 /* fixme arbitrary */
+static void ocpp_wsreader(void *p1, void *p2, void *p3)
+{
+	struct ocpp_info *ctx = p1;
+	struct ocpp_upstream_info *ui = ctx->ui;
+	struct zsock_pollfd tcpfd;
+	uint8_t ret;
+	uint8_t retry_cnt = 0;
+
+	ctx->is_cs_offline = true;
+	tcpfd.fd = ui->tcpsock;
+	tcpfd.events = ZSOCK_POLLIN | ZSOCK_POLLERR
+			| ZSOCK_POLLHUP | ZSOCK_POLLNVAL;
+
+	while (1) {
+
+		if (ctx->is_cs_offline &&
+		    !(retry_cnt++ % TCP_CONNECT_AFTER)) {
+
+			k_mutex_lock(&ctx->ilock, K_FOREVER);
+			ocpp_connect_to_cs(ctx);
+			k_mutex_unlock(&ctx->ilock);
+		}
+
+		ret = zsock_poll(&tcpfd, 1, 200);
+		if (ret <= 0) {
+			continue;
+		}
+
+		if ((tcpfd.revents & ZSOCK_POLLERR) ||
+		    (tcpfd.revents & ZSOCK_POLLHUP) ||
+		    (tcpfd.revents & ZSOCK_POLLNVAL)) {
+			LOG_ERR("poll err %d", tcpfd.revents);
+			ctx->is_cs_offline = true;
+			continue;
+		}
+
+		if (tcpfd.revents & ZSOCK_POLLIN) {
+			struct ocpp_wamp_rpc_msg rcv = {0};
+
+			rcv.ctx = ctx;
+			rcv.msg = ui->recv_buf;
+			rcv.sndlock = &ui->ws_sndlock;
+			rcv.msg_len = CONFIG_OCPP_RECV_BUFFER_SIZE;
+
+			memset(ui->recv_buf, 0, rcv.msg_len);
+			ret = ocpp_receive_from_server(&rcv, 200);
+			if (ret < 0) {
+				continue;
+			}
+
+			ocpp_process_server_msg(ctx);
+		}
+	}
+}
+
+int ocpp_upstream_init(struct ocpp_info *ctx, struct ocpp_cs_info *csi)
+{
+	int ret;
+	struct ocpp_upstream_info *ui;
+
+	LOG_INF("upstream init");
+
+	ui = k_calloc(1, sizeof(struct ocpp_upstream_info));
```

Comment:
You probably have reasons for doing dynamic memory allocations but we have been trying to avoid these (which has not always been possible) in networking code. If you do not do dynamic memory allocations you cannot have memory leaks. You could still have resource leaks but that is another issue.
Anyway, not a deal breaker but please rethink would it be possible to avoid using malloc and friends.